core/src/main/scala/org/apache/spark/executor/ExecutorMetrics.scala

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.executor
+
+import org.apache.spark.annotation.DeveloperApi
+
+/**
+ * :: DeveloperApi ::
+ * Metrics tracked during the execution of an executor.
+ *
+ * So, when adding new fields, take into consideration that the whole object can be serialized for
+ * shipping off at any time to consumers of the SparkListener interface.
+ */
+@DeveloperApi
+class ExecutorMetrics extends Serializable {
+
+  type SysTime = Long
+
+  /**
+   * Host's name the executor runs on
+   */
+  private var _hostname: String = _
+  def hostname: String = _hostname
+  private[spark] def setHostname(value: String) = _hostname = value
+
+  /**
+   * Host's port the executor runs on
+   */
+  private var _port: Int = _
+  def port: Int = _port
+  private[spark] def setPort(value: Int) = _port = value
+
+  def hostPort: String = hostname + ":" + port
+
+  /**
+   * maximum on-heap memory that the executor used for shuffle read on Netty network layer
+   */
+  @volatile private var _maxNettyReadOnheapSizeTime: (Long, SysTime) = (0L, 0L)
+  def maxNettyReadOnheapSizeTime: (Long, SysTime) = _maxNettyReadOnheapSizeTime
```

Comment:
I think I agree with @jerryshao .

(a) the time here is just a timestamp that we got the max, it doesn't make sense to take a sum of them, and the case class will make it more clear.  but instead of `currentTime` I'd use `timestamp`

(b) I'm not sure if there is any value in including "read" in the name.  its true that netty is mostly using this memory for reading data off the network, but we're not actually doing anything special to make sure we only count memory used during read.  We are just counting the total memory used by netty, and it so happens that the write-side is streaming.  (I think I may have been confused during earlier discussions -- the write-side uses `FileSegmentManagedBuffer.convertToNetty` which should by streaming.)  We should mention in the docs that the primary source of this memory is from the shuffle-read, that netty will needs memory for the blocks fetched from each executor thread, etc. But I think we can only guarantee that we get statistics on netty's memory usage in general, we can't necessarily separate out the write-side.